### PR TITLE
Apiv2service

### PR DESCRIPTION
**Pull request by baude**
Add podman service command that allows users to run either a RESTful or
varlink protocol API service.

Addition of docs and RESTful listening.

### CODE REVIEW

**Comment by mheon** [comment]
Can we get better descriptions on these? The API blogpost is out, and it really helps reviewers.

File: docs/source/markdown/podman-service.1.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+% podman-search(1)
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-service.1.md]: s/service/search

File: docs/source/markdown/podman-service.1.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## NAME
+podman\-service - Run an API service
+
+## SYNOPSIS
+**podman service** [*options*]
+
+## DESCRIPTION
+The **podman service** command creates a listening service that will answer API calls for podman.  You may
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-service.1.md]: I'd pref "API calls for Podman"

File: docs/source/markdown/podman-service.1.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+optionally provide an endpoint for the API in URI form.  For example, *unix://tmp/foobar.sock* or *tcp:localhost:8080*.
+If no endpoint is provided, defaults will be used.  The default endpoint for a rootfull
+service is *unix:/run/podman/podman.sock* and rootless is *unix:/$XDG_DIR/podman/podman.sock* (for
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-service.1.md]: XDG_DIR or XDG_RUNTIME_DIR?

File: docs/source/markdown/podman-service.1.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+Use the varlink protocol instead of the REST-based protocol.  This option will be deprecated in the future.
+
+**--help**, **-h**
+
+Print usage statement
+
+## EXAMPLES
+
+Run an API listening for 5 seconds using the default socket
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-service.1.md]: ending period

File: docs/source/markdown/podman-service.1.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+The time until the session expires in _milliseconds_. The default is 1
+second. A value of `0` means no timeout and the session will not expire.
+
+**--varlink**
+
+Use the varlink protocol instead of the REST-based protocol.  This option will be deprecated in the future.
+
+**--help**, **-h**
+
+Print usage statement
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-service.1.md]: ending period

File: docs/source/markdown/podman-service.1.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+## EXAMPLES
+
+Run an API listening for 5 seconds using the default socket
+```
+podman service --timeout 5000
+```
+
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-service.1.md]: did you want this extra blank line?

File: pkg/api/server/server.go [hunk 7]
```diff
@@ -48,24 +48,47 @@ import (
...
 	net.Listener       // mux for routing HTTP API calls to libpod routines
 	context.CancelFunc // Stop APIServer
 	*time.Timer        // Hold timer for sliding window
 	time.Duration      // Duration of client access sliding window
 }
 
-// NewServer will create and configure a new API HTTP server
+// Number of seconds to wait for next request, if exceeded shutdown server
+const (
+	DefaultServiceDuration   = 300 * time.Second
```
Comment (TomSweeneyRedHat) [inline pkg/api/server/server.go]: any thoughts about adding this to the conf file?  5 minutes seems very reasonable though in probably 99.999999999999% of the cases.

Comment (baude) [reply pkg/api/server/server.go]: it is the default, but is overriden by user with an option.

File: docs/source/markdown/podman-service.1.md (hunk 3 above)
Comment (baude) [reply docs/source/markdown/podman-service.1.md]: i think i have it correct ... anyone else?

**Comment by baude** [comment]
@rhatdan ptal

**Comment by baude** [comment]
fwiw, need this to be able to write tests ... i know @edsantiago is looking for something similar. I need for go-binding tests

File: cmd/podman/service.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	if rootless.IsRootless() {
+		xdg, err := util.GetRuntimeDir()
+		if err != nil {
+			return err
+		}
+		socketName := "podman.sock"
+		if c.Varlink {
+			socketName = "io.podman"
+		}
+		socketDir := filepath.Join(xdg, "podman", socketName)
```
Comment (jwhonce) [inline cmd/podman/service.go]: Does this need to be a secure join to prevent nastiness coming from the xdg variable?

File: cmd/podman/service.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+		if _, err := os.Stat(filepath.Dir(socketDir)); os.IsNotExist(err) {
+			if err := os.Mkdir(filepath.Dir(socketDir), 0755); err != nil {
+				return err
+			}
+		}
+		apiURI = fmt.Sprintf("unix:%s", socketDir)
```
Comment (jwhonce) [inline cmd/podman/service.go]: You are ignoring err's that are not os.IsNotExist()

File: cmd/podman/service.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+	timeout := time.Duration(c.Timeout) * time.Millisecond
+	if c.Varlink {
+		return runVarlink(runtime, apiURI, timeout, c)
+	}
+	return runREST(runtime, apiURI, timeout)
+}
+
+func runREST(r *libpod.Runtime, uri string, timeout time.Duration) error {
+	fmt.Println("This function is EXPERIMENTAL.")
```
Comment (jwhonce) [inline cmd/podman/service.go]: Should this also be logged?

File: docs/source/markdown/podman-service.1.md (hunk 3 above)
Comment (jwhonce) [reply docs/source/markdown/podman-service.1.md]: @baude Tom is correct. XDG_RUNTIME_DIR

**Pull request merged**